crates/bdk/src/wallet/mod.rs

```diff
@@ -2312,6 +2312,55 @@ impl<D> Wallet<D> {
     pub fn local_chain(&self) -> &LocalChain {
         &self.chain
     }
+
+    /// Set lookahead for all keychains.
+    ///
+    /// Passing a lookahead of 0 will result in an error. This is because if we sync
+    /// the chain with a lookahead of 0, we would not find any update since we don't
+    /// have any scripts stored.
+    pub fn set_lookahead_for_all(&mut self, lookahead: u32) -> Result<(), Error> {
+        if lookahead == 0 {
+            return Err(Error::Generic(
+                "lookahead must be greater than 0".to_string(),
+            ));
+        }
+        self.indexed_graph.index.set_lookahead_for_all(lookahead);
+        Ok(())
+    }
+
+    /// Insert all the block's relevant transactions into the IndexedTxGraph.
+    pub fn apply_block_relevant(
+        &mut self,
+        block: Block,
+        height: u32,
+    ) -> Result<(), CannotConnectError>
+    where
+        D: PersistBackend<ChangeSet>,
+    {
+        let chain_update = CheckPoint::from_header(&block.header, height).into_update(false);
+        let mut changeset = ChangeSet::from(self.chain.apply_update(chain_update)?);
+        changeset.append(ChangeSet::from(
+            self.indexed_graph.apply_block_relevant(block, height),
```

Comment:
Not saying this should be changed, but worth pointing out: internally `apply_block_relevant` will use `apply_batch_relevant`. The latter can be used on any batch and therefore can't make any assumption on the topology, for this reason it loops twice on the list of transactions (once for updating the spk index, once for updating the tx graph). In the case of valid blocks, where transactions must be topologically ordered, we can just use a single loop.

Not a huge optimization, but still, avoids a few thousands iterations.

Reply:
Sure. Worth a comment that we could make an assumption on the topology of the order of the transactions to make a bespoke implementation.

Reply:
Implemented in d55a40799428754d9e33302f9c6057287b8a759f